File: src/test/java/nl/vollo/kern/api/GebruikerCtlApiTest.java
package nl.vollo.kern.api;

import nl.vollo.kern.test.ApiTestService;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import java.io.IOException;

import static nl.vollo.kern.test.JsonNodeIsEqualToTextMatcher.isTextNode;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

@ExtendWith(SpringExtension.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@DisplayName("GebruikerCtlApi")
class GebruikerCtlApiTest {

    @LocalServerPort
    private int port;

    @Autowired
    private TestRestTemplate restTemplate;

    @Autowired
    private ApiTestService apiTestService;

    @BeforeAll
    void beforeAll() {
        apiTestService.init(port, restTemplate).genererenTestdata();
    }

    @Test
    @DisplayName("geeft ingelogde gebruiker terug na inloggen")
    void inloggen() throws IOException {
        var headers = apiTestService.inloggen();
        var response = restTemplate.exchange(
                apiTestService.testUrl("/gebruiker/ingelogd"), HttpMethod.GET,
                new HttpEntity<Void>(headers), String.class);

        assertThat(response.getStatusCode(), is(HttpStatus.OK));
        var json = apiTestService.getObjectMapper().readTree(response.getBody());
        assertThat(json.get("gebruikersnaam"), isTextNode("m0"));
        assertThat(json.get("rollen"), isTextNode("ROLE_GEBRUIKER"));
        assertThat(json.get("_type"), isTextNode("GEBRUIKER"));
        assertThat(json.get("medewerker").get("voornaam"), isTextNode("Kwaku"));
        assertThat(json.get("medewerker").get("achternaam"), isTextNode("Oudejans"));
        assertThat(json.get("medewerker").get("_type"), isTextNode("MEDEWERKER"));
    }

    @Test
    @DisplayName("geeft geen ingelogde gebruiker indien niet ingelogd")
    void nietIngelogd() {
        var response = restTemplate.exchange(
                apiTestService.testUrl("/gebruiker/ingelogd"), HttpMethod.GET, null, String.class);

        assertThat(response.getStatusCode(), is(HttpStatus.UNAUTHORIZED));
    }
}

File: src/test/kotlin/nl/vollo/kern/api/GebruikerCtlApiTest.kt
package nl.vollo.kern.api

import com.fasterxml.jackson.databind.ObjectMapper
import nl.vollo.kern.test.isTextNode
import nl.vollo.kern.testdata.TestdataGenerator
import org.hamcrest.CoreMatchers
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.test.web.client.exchange
import org.springframework.boot.test.web.client.postForEntity
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.http.*
import org.springframework.test.context.junit.jupiter.SpringExtension

@ExtendWith(SpringExtension::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
internal class GebruikerCtlApiTest {

    @LocalServerPort
    private var port: Int = 0

    private val objectMapper = ObjectMapper()

    @Autowired
    private lateinit var restTemplate: TestRestTemplate

    @Autowired
    private lateinit var testdataGenerator: TestdataGenerator

    @BeforeAll
    fun beforeAll() {
        testdataGenerator.genereren(
                aantalScholen = 1,
                aantalGroepen = 1,
                aantalLeerlingenMin = 3,
                aantalLeerlingenMax = 5
        )
    }

    @Test
    fun getIngelogdeGebruiker() {
        val token = inloggen()
        val headers = HttpHeaders().apply { set("Cookie", token) }
        val gebruiker: ResponseEntity<String> = restTemplate.exchange(testUrl("/gebruiker/ingelogd"), HttpMethod.GET,
                HttpEntity<Void>(headers))
        assertThat(gebruiker.statusCode, equalTo(HttpStatus.OK))
        println(gebruiker.body)
        val json = objectMapper.readTree(gebruiker.body)
        assertThat(json["gebruikersnaam"], isTextNode("m0"))
        assertThat(json["rollen"], isTextNode("ROLE_GEBRUIKER"))
        assertThat(json["_type"], isTextNode("GEBRUIKER"))
        assertThat(json["medewerker"]["voornaam"], isTextNode("Gerharda"))
        assertThat(json["medewerker"]["achternaam"], isTextNode("Hoef"))
        assertThat(json["medewerker"]["_type"], isTextNode("MEDEWERKER"))
    }

    @Test
    @Disabled
    fun logout() {
    }

    fun inloggen(): String {
        val inloggenRequest = InloggenCtl.InloggenRequest("m0", "m0")
        val result: ResponseEntity<Void> = restTemplate.postForEntity(testUrl("/public/inloggen"), inloggenRequest)
        assertThat("Fout bij inloggen", result.statusCode, equalTo(HttpStatus.NO_CONTENT))
        val cookie: String? = result.headers.getFirst("Set-Cookie")
        assertThat("Geen cookie ontvangen bij inloggen", cookie, CoreMatchers.notNullValue())
        assertThat("Geen token ontvangen bij inloggen", cookie!!.startsWith("vollo_token"), CoreMatchers.equalTo(true))
        return cookie
    }

    fun testUrl(path: String) = "http://localhost:$port$path"
}